### PR TITLE
app.go: Take a pass over the godoc

### PR DESCRIPTION
**Pull request by abhinav**
The godocs for most of the public API in app.go hasn't been touched in a
while.
I took a quick pass over it to adjust it better to current
recommendations and best practices.
This also makes more use of Go's `[..]` linking.

One notable change: Instead of Options, logical groups of options are
now recomended via Module.

### CODE REVIEW

File: annotated.go [hunk 1]
```diff
@@ -109,8 +111,9 @@ var (
 	}
 )
 
-// Annotation can be passed to Annotate(f interface{}, anns ...Annotation)
-// for annotating the parameter and result types of a function.
+// Annotation decorates a target for [Annotate].
```
Comment (JacobOaks) [inline annotated.go:114]: This overloads "decorate" a little bit. Consider a different word choice.

File: app.go [hunk 2]
```diff
@@ -125,6 +111,10 @@ func (og optionGroup) String() string {
 }
 
 // StartTimeout changes the application's start timeout.
+// This controls the total time that all OnStart hooks have to complete.
```
Comment (JacobOaks) [inline app.go:114]: I think linking to this here would provide some value.
```suggestion
// This controls the total time that all [OnStart] hooks have to complete.
```

File: app.go [hunk 3]
```diff
@@ -145,6 +135,10 @@ func (t startTimeoutOption) String() string {
 }
 
 // StopTimeout changes the application's stop timeout.
+// This controls the total time that all OnStop hooks have to complete.
```
Comment (JacobOaks) [inline app.go:138]:
```suggestion
// This controls the total time that all [OnStop] hooks have to complete.
```

File: app.go [hunk 4]
```diff
@@ -186,18 +180,27 @@ func (o recoverFromPanicsOption) String() string {
...
 // For example,
 //
 //	WithLogger(func(logger *zap.Logger) fxevent.Logger {
 //	  return &fxevent.ZapLogger{Logger: logger}
 //	})
+//
+// If specified, Fx will construct the logger and log all its events to the
+// specified logger.
+//
+// If Fx fails to build the logger, it will fall back to
```
Comment (JacobOaks) [inline app.go:202]:
```suggestion
// If Fx fails to build the logger, or no logger is specified, it will fall back to
```

File: app.go [hunk 5]
```diff
@@ -186,18 +180,27 @@ func (o recoverFromPanicsOption) String() string {
 	return "fx.RecoverFromPanics()"
 }
 
-// WithLogger specifies how Fx should build an fxevent.Logger to log its events
-// to. The argument must be a constructor with one of the following return
-// types.
+// WithLogger specifies the logger used by Fx to log its own events
```
Comment (JacobOaks) [inline app.go:183]:
```suggestion
// WithLogger specifies the [fxevent.Logger] used by Fx to log its own events
```

File: app.go [hunk 6]
```diff
@@ -224,14 +227,15 @@ func (l withLoggerOption) String() string {
 // Printer is the interface required by Fx's logging backend. It's implemented
 // by most loggers, including the one bundled with the standard library.
 //
-// Note, this will be deprecate with next release and you will need to implement
-// fxevent.Logger interface instead.
+// Note, this will be deprecate in a future release.
```
Comment (JacobOaks) [inline app.go:230]:
```suggestion
// Note, this will be deprecated in a future release.
```

**Review by JacobOaks — APPROVED** [review]
Have a couple suggestions, but thanks this looks great!

**Review by sywhang — APPROVED** [review]

**Pull request merged**